### ISSUE TITLE
Create tutorials and documentation meant for developers that wish to build on the Sphinx stack

### ISSUE THREAD

**Issue by choldgraf** [executablebooks/meta#837]
### Context

The Python implementation of MyST and Jupyter Book depends heavily on Sphinx, and the ways to extend Jupyter Book will require an understanding of how Sphinx works. However, Sphinx and docutils are both very complex, and both of them lack complete and accessible documentation. This makes it hard for people to learn about how our projects work and how to contribute, and also makes it harder for them to extend and build on top of the stack.

### Proposal

We should have a dedicated space in the documentation to help people in a few ways to make it easier for them to learn and contribute. Here are a few that came to mind immediately:

- get started with the docutils/sphinx stack
- understand how Jupyter Book and its components use that stack
- pointers to take first-steps towards extending and building on that stack.

**Comment by kai-tub** [comment]
Thanks for opening an issue!
I know that this is a slippery slope, but I really think that there are some areas where 'quick' tutorials would really help external developers. Sphinx alone is a huge beast that took me forever to get the basics running, and I still have no clear idea how to test it.

I guess my point is that I know that it seems like a never-ending task to write tutorials about so many tools, but if I can get away with looking at a short mini-tutorial that links to the relevant source code section for more information, I am already a lot better off!
I personally don't even mind looking at the source code but with so many projects, it is sometimes hard to track what repositories are even relevant.

I am quite sure that everybody is aware of these projects, but IMHO these are projects with the best documentation about their "stack":
- [FastAPI](https://fastapi.tiangolo.com/) or [SQLModel](https://sqlmodel.tiangolo.com/) (or anything from  [Sebastián](https://github.com/tiangolo) really)
- [furo](https://pradyunsg.me/furo/) for its clear navigation (see customisation section)
- And not yet out, but still on many people's radar: [textual](https://github.com/Textualize/textual/tree/css)
- .... Ok, I am blanking. There are a couple of other good examples for an architectural overview, but those are the 'heavy' hitters IMHO 

PS: Slightly off-topic. I also think that [this article from Sebastian](https://dev.to/tiangolo/the-future-of-education-and-art-4h0e) about the future of education and art could be nicely targeted towards the executable book audience. Especially with many users in educational roles, maybe this could become a platform to invite "artists for education"?

**Comment by chrisjsewell** [comment]
Heya, I would suggest this needs to be decided as an off-shoot of high-level strategy (see e.g. #839)

In particular, should we be encouraging users to work with in the sphinx/docutils space, if (potentially) the high-level strategy is to move to a JavaScript implementation (see e.g. #838)?
Naturally, then any work by developers or users (making e.g. sphinx extensions) within this space, could be an inefficient use of time and resources.

We should also make sure we are not duplicating upstream efforts and/or trying to upstream any work we do. There is already some good work done, primarily by the RTD guys, to add sphinx tutorials: https://github.com/sphinx-doc/sphinx/issues/9165

**Comment by NickleDave** [comment]
Hope it's okay that I chime in as an outsider.  
I totally get the reasons for moving to js (I think). And so it makes sense to not develop docs that would then take time and energy from the main thing you are working on.
  
But there's a world of Python users that do not have any of the insight you all have built up from working with docutils and sphinx. When I heard @choldgraf @pradyunsg  et al. on Talk Python describing the internals, I was really wishing that I could read about that somewhere.  

And clearly there are still people in Python land thinking about the tooling 
https://github.com/sphinx-doc/sphinx/issues/8039
https://github.com/reStructuredText/startup/issues/1

Would it be worth writing some summary of "lessons learned about Sphinx/docutils internals" with a big disclaimer at the top that says "we are focusing on js for reasons x, y, z, please see these other doc pages/blog posts/etc". Like, a post mortem (you can give it a more polite title).

Of course in general reduplication of effort is bad. But as you all know better than me there's a bit of a divide between Jupyter users and contributors because of javascript. If you at least write down what you learned in a way that's a little more readable for the general public, maybe other Python devs could develop complementary tooling for all the users that aren't going to master Typescript

I'd love to see a post with diagrams like in this presentation
https://docs.google.com/presentation/d/168yre5u_D2wQpeySrrDqV3cM9qE85YiaRTT8tMpjcGo/htmlpresent
that are not trapped inside an html-published slide deck from 2015 that no one can download easily,
maybe using the awesome mermaid directive that some brilliant EBP developers came up with :grin:

**Comment by chrisjsewell** [comment]
Heya thanks @NickleDave 

> But as you all know better than me there's a bit of a divide between Jupyter users and contributors because of javascript

Yeh I think that's the same kind of here, and more generally its just this trade-off that javascript is not particularly nice to use (Python is certainly better) but it has been the only game in town for web/browser-based development.
Now we have e.g. WASM (-> pyiodide etc) but still all the tooling is not yet there to really use this to replace JS.

So thats why JS.

Then even if we could use Python, the problem with Sphinx, is that what we really want is a "reference implementation" for https://github.com/executablebooks/myst-spec, and its basically impossible to do this on top of a framework that we don't control, and certainly not one as messy and complicated as sphinx.
(Sphinx as a framework also has some quite "severe" limitations, when it comes to building something for the browser.)

Sphinx does still have a certain degree of method to its madness though lol, I for instance added the order of events list here: https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx-core-events, and it is possible to "master"

**Comment by chrisjsewell** [comment]
> et al. on Talk Python describing the internals

yeh I guess I was the et al here lol

In fact you want a lessons learned, here's one I made earlier ...

---

# MyST Parser Design

In this document we outline the general design decisions for a generic MyST parser, and then how this applies to the Javascript parser we have built here.
Note, this may eventually be moved to a "top-level" documentation of MyST.

Currently, the primary implementation of a MyST parser is written as a Sphinx extension (in Python); using markdown-it to initially parse the source text to a "token stream" (a list of syntax tokens, encapsulating the whole document and its content), then we convert this token stream to a docutils AST tree (in the `myst-parser` extension), which Sphinx then uses to convert to the desired output format (e.g. HTML or LaTeX).
Naturally this design is tightly coupled to Sphinx, but (a) in Javascript we do not have an implementation of Sphinx, and (b) we would like to move away from being reliant on any one "technology" for parsing, and instead outline a more generic "standard" for MyST parsing, which anyone could in principle implement.
What we don't want to do though is end up unknowingly reimplementing a worse version of Sphinx.
In the next section then we discuss the Sphinx design, the reasons behind it, and some of its technical limitations.

## Analysis of the Sphinx design

The sphinx design is outlined in more detail at <https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx-core-events>, but the basic stages can be described as:

1. We read in a global configuration for the parse.
2. We need to parse the each document into an "output format agnostic" Abstract Syntax Tree (AST). This is performed in a linear manner, stepping through each line of the source text.
   - As well as creating the AST here, we also store aspects of the document to a global state object (known as the `BuildEnvironment`), for later fast lookup.
   - Also to note here, directive and role syntaxes are processed as they are encountered.
3. There are certain per-document AST operations we cannot perform until we have parsed that document, e.g. replacing substitution references with their definitions. These are known as transforms, and are applied in order of priority.
4. We then want to cache each document AST, so that we do not have to re-parse every document when one changes.
   - We also need to cache the global state object
5. There are certain per-document AST operations we cannot perform until we have parsed all documents, i.e our global state is complete and up-to-date, for example matching inter-document references to their targets. These are known as post-transforms, and are applied in order of priority.
   - Additionally, we may want to perform operations specific to a certain output format. These are also included in the post-transforms, and so the post-transforms are run once per each output format.
   - The per document ASTs at this point are transient, i.e. they are not cached, since any change to any document could affect them, and so there is no benefit in caching.
6. Now we have our final ASTs, we can perform the render, whereby we convert the ASTs to the output format.
   - A renderer in sphinx is known as a `Builder`
   - Another important thing we need to do is map filepath references (e.g. for images and downloadable files) to paths in the build folder, and ensure these files are copies there.

Another core concept is that of the logger, which logs specific information/warnings to the console, but also can be configured to fail the build (i.e. produce a non-zero exit code) if any warnings are encountered.
In this way the build is robust to errors (we don't want the whole build failing because of one syntax error), but allows us to programmatically tell if there any issues with our documentation (e.g. when we run CI tests).

As an addendum to the above design, we can also consider the steps to re-build the outputs, given an initial build has already been performed.

1. Within the global configuration specification, each variable defines a rebuild condition, i.e. whether a change in this variable should invoke a full rebuild (invalidating the cached document ASTs and global env and starting again from step (2)) or simply requires a rebuild from step (5).
   - The variables from the last parse are stored in the global env, and compared to those from the current parse.
2. In deciding which documents should be reparsed from step (2), the mtime of the source file and cached AST file are compared
3. step (5) and (6) are always run.

Lastly we should consider Sphinx's plugin system, in the form of extensions which can:

1. Specify additional configuration variables (including type validation and rebuild condition)
2. Define functions that occur after the configuration has been read (known as `config-inited` events) e.g. to apply additional validation
3. Add new parsers to apply to particular file name suffixes
4. Define additional roles and directives
5. Define additional transforms (and their priority)
6. Define additional post-transforms (their priority and what output formats they apply to)
7. Define additional renderers
8. Override which cached documents are considered outdated/invalid (known as `env-get-outdated` events)
9. Interject at a number of other key stages in the build (see other events)

Although a lot of this system is well designed, and we will certainly need to include most if not all of these steps, there a number of design issues that could be improved:

- The configuration must be written as Python file, instead it should be a more general format (like YAML or TOML)
- The document ASTs are designed as Python class instances (see <https://github.com/chrisjsewell/docutils/blob/develop/docutils/docutils/nodes.py>) and cached as pickled files (see https://docs.python.org/3/library/pickle.html). Again this should be a more general format such a JSON.
  - Usually when testing docutils ASTs we use the `pformat` method which converts it into a "pseudo-XML" string, although this does not actually contain all the information about the AST.
  - The markdown-it token stream system allows for a much more facile serialisation/de-serialisation to JSON
- The environment object is again a Python class instance and stored by pickling.
  - More so than just having a general storage format here this is currently a big obstacle for programs that want to access the environment in a dynamic manner, for example language servers that could provide auto-completions and navigation for inter-file references. Here it would be really useful for the environment to be a database with concurrent read/write access.
  - <https://en.wikipedia.org/wiki/SQLite> would possibly be a good choice here, as a serverless DB format (i.e. does not require a background process to be running), with support for basically every programming language.
  - See <https://github.com/chrisjsewell/rst-language-server>, where I was working on such a language server and had to adapt docutils/sphinx in such a way (it was also beneficial to store the line number mappings for references etc)
- Mappings of directive/role names to their processing functions/classes are stored in docutils as global variables. This is problematic for asynchronous document parsing.
- The conversion of roles/directives is performed in the same pass as with all other syntaxes. This leads to the AST being intrinsically "lossy" in that there is no way to recover what directives were in the original source text (see <https://github.com/executablebooks/rst-to-myst> where I had to re-write some of the parsing code to achieve this).
- Role/directive/transform code is very "side-effect heavy", as in they directly mutate the document AST and environment. There is already plenty of literature out there explaining why this is not ideal (e.g. <https://softwareengineering.stackexchange.com/questions/15269/why-are-side-effects-considered-evil-in-functional-programming>), and in general we should strive towards a more functional programming paradigm.
- Overall sphinx is quite difficult to run in a programmatic manner (as opposed to via the command-line)
- Sphinx is quite tied to having files exist on a file system

**Comment by NickleDave** [comment]
> So thats why JS.

All makes sense to me

> yeh I guess I was the et al here lol  

In fact you were, sorry! Hadn't made the connection, I blame common names like "Chris". Almost as bad as "David" :roll_eyes: :smirk_cat: 

Thank you for taking the time to share the links to the order of execution and what you had written up previously, both are definitely informative to me. 

I can totally understand not wanting to sound like you are criticizing the really hard-working Sphinx and docutils devs/maintainers who have their own project goals/histories/trade-offs, but something like this as a blog post I do think could be of wide interest 

> Sphinx is quite tied to having files exist on a file system  

Can you say a little more about this point? I saw it in other places too.  
I think what's implied here is that processing steps fail when files don't exist? And the alternative would be to have an intermediate step that works with e.g., strings, so one can do something like `load`, `loads`, `dump`, `dumps`?

**Comment by chrisjsewell** [comment]
Thanks!

> I blame common names like "Chris". Almost as bad as "David"

Indeed 😆 

> Sphinx is quite tied to having files exist on a file system
> Can you say a little more about this point? I saw it in other places too.

Yeh so Ideally, I would say, you want some kind of interface layer, whereby you could supply a "virtual file system" for sphinx to work on, using e.g. the python `pathlib.Path` API
In fact as an example I would give my own https://github.com/aiidateam/archive-path, whereby you could have sphinx read a zip file

**Comment by kai-tub** [comment]
@chrisjsewell thank you for providing such a complete overview of your experience with Sphinx/Docutils!
I agree with @NickleDave that it would be quite helpful to have this 'formalized' into a blog post to point others toward.

**Comment by choldgraf** [comment]
Hey all - just a quick note that I provide some explanation of my perspective on the Jupyter Book / MyST JS differentiation and plan here: 

- https://github.com/executablebooks/meta/issues/838#issuecomment-1273076387

For this issue of tutorials etc, I agree completely that we should offload as much team knowledge about Sphinx and Docutils as possible into the broader community. In my opinion this is where we should direct a significant amount of our capacity on the current grant, as it is crucial to spread knowledge of this stack across multiple people.

I'm indifferent as to whether we do this in an Executable Books-specific place, or via upstream contributions to other places. I think the most important thing is to _get the information out there_ because this is where the community is currently bottlenecked. We should take whatever path has the least resistance for now, and decide what to do with the information later.

**Comment by kai-tub** [comment]
I would like to follow up on this a bit.
While reading various independent discussions in this repository, it is quite hard to get motivated to dive deeper than _necessary_ into Sphinx (from a developer's perspective), as it seems like the Sphinx ecosystem is quite unapproachable and doesn't look like it will be the _future_ of the MyST ecosystem.
Though, I still think it would be nice to have some of your knowledge put into short tutorials.

I think that there is almost no question that the executable books project belongs to one of the most experienced "outside" sphinx users/developers (maybe with the readthedocs team) and that you are quite familiar with its shortcomings and limitations.

I am aware that this has been discussed in a Sphinx issue (https://github.com/sphinx-doc/sphinx/issues/9165) and that there has been some work to onboard new users, but I am still missing a good introduction for testing Sphinx.
From my perspective, this seems to be the most elusive component and is only solvable by searching through various GitHub issues and reading through code (and this still seems to be quite unhelpful as Sphinx apparently has "special" handling for their own test suite?).

I know that from a testing perspective, the MyST ecosystem had to fight quite a bit with directive parsing and post-transforms, but even a trivial 'CLI' guided testing section would be tremendously helpful for writing 'simple' tests.
Again, I am aware that we probably can all agree that it would be _nice_ if there were a string-based (non-disk file-focused) API, but given that it looks like this will be a task for a different documentation framework....
I would really appreciate it if the ecosystem of the executable book tries to upstream their "testing" knowledge for the "Sphinx way" (a tutorial and a couple of quick how-to guides a la https://documentation.divio.com/introduction/; PS: Just to show you that I try to read all of your posts across all projects :D ) and a discussion/review post that describes the shortcomings of the current implementation in a separate blog post.
I assume that your idea would be to write a 'comparison' with the custom JS implementation vs. Sphinx and discuss the various trade-offs, but I would argue that it would already be quite handy to have a general/conceptual discussion on what a 'new' framework could learn from Sphinx. Especially as this would be quite interesting for people that are interested in the 'abstract' knowledge and maybe not necessarily in the implementation details of Sphinx.
Again, I know that this knowledge is somewhat already present in various issue discussions, but it is quite inaccessible and hard to find for later review.

TL;DR: IMHO, I think it would be nice if the executable books project could fill in a couple of holes in the Sphinx documentation regarding testing, as I think this project has the most in-depth knowledge on this subject aside from the core maintainers. I know that the testing methodology of Sphinx is maybe not be how many would like it to be, but a general tutorial for others that will have to use Sphinx would be very helpful for the wider community. Also, given the experience, it would also be nice to have a single blog post, accumulating the experience and giving a single reference to what an alternative solution could do differently.
